addons/base_import_module/models/ir_module.py

```diff
@@ -258,6 +268,125 @@ def module_uninstall(self):
+        return {
+            'name': _('Apps'),
+            'type': 'ir.actions.act_window',
+            'res_model': 'ir.module.module',
+            'view_mode': 'form',
+            'res_id': self.id,
+            'context': self.env.context,
+        }
+
+    def web_read(self, specification):
+        fields = list(specification.keys())
+        module_type = self.env.context.get('module_type', 'official')
+        if module_type != 'official':
+            modules_list = self._get_modules_from_apps(fields, module_type, self.env.context.get('module_name'))
+            return modules_list
+        else:
+            return super().web_read(specification)
+
+    @api.model
+    def _get_modules_from_apps(self, fields, module_type, module_name):
+        payload = {
+            'series': major_version,
+            'module_fields': fields,
+            'module_type':module_type,
+            'module_name': module_name
+        }
+        headers = {'Content-Type': 'application/json'}
+        resp = requests.post(
+            f"{APPS_URL}/loempia/listdatamodules",
+            data=dumps({'params': payload}),
+            headers=headers,
+        )
+        modules_list = resp.json().get('result', [])
+        for mod in modules_list:
+            mod['id'] = -1
+            module_name = mod.get('name', module_name)
+            if 'icon' in fields:
+                mod['icon'] = f"{APPS_URL}{mod['icon']}"
+            if 'state' in fields:
+                existing_mod = self.search([('name', '=', module_name), ('state', '=', 'installed')])
+                if existing_mod:
+                    mod['state'] = 'installed'
+                else:
+                    mod['state'] = 'uninstalled'
+            if 'module_type' in fields:
+                mod['module_type'] = module_type
+            if 'website' in fields:
+                mod['website'] = f"{APPS_URL}/apps/modules/{major_version}/{module_name}/"
+            if 'app_iframe' in fields:
+                mod['app_iframe'] = '<iframe src="%s/?no_header=1" title="Apps Page"></iframe>' % mod['website']
+        return modules_list
+
+    def button_immediate_install_app(self):
+        module_name = self.env.context.get('module_name')
+        resp = requests.get(
+            f"{APPS_URL}/loempia/download/data_app/{module_name}/{major_version}"
+        )
+        import_module = self.env['base.import.module'].create({
+            'module_file': base64.b64encode(resp.content),
+            'state': 'init',
+            'modules_dependencies': self._get_missing_dependencies(resp.content)
+        })
+
+        return {
+            'name': 'Install an App',
+            'view_mode': 'form',
+            'target': 'new',
+            'res_id': import_module.id,
+            'res_model': 'base.import.module',
+            'type': 'ir.actions.act_window',
+            'context': {'data_module': True}
+        }
+
+    @api.model
+    def _get_missing_dependencies(self, zip_data):
```

Comment:
isn't it a bit overkill to unzip and parse the module just to read a manifest key? couldn't we have that info from apps instead?

Reply:
Also it's a lot of copy/paste from
https://github.com/odoo/odoo/pull/140245/files#diff-3d414e7f4df2b7703572278aa6906de5dec69481842c0ea364eca2e20e6f6e3dR196-R216

Maybe there is a way to factorize this to avoid this copy/paste

Reply:
In the following fixup:https://github.com/odoo/odoo/pull/140245/commits/c419521087e54ac223b56b5c01da947b080f215e I just read the manifest files (from memory)

It seems hard to factorize with the other function that needs to  extract all files from the zip, that may contain bigger files than manifest (like images)

If you have any other comments, they are welcome.